client/allocrunner/consul_hook.go

```diff
@@ -145,29 +171,60 @@ func (h *consulHook) prepareConsulTokensForTask(task *structs.Task, tg *structs.
 
 	// Find signed workload identity.
 	ti := *task.IdentityHandle(wid)
-	jwt, err := h.widmgr.Get(ti)
+	swi, err := h.widmgr.Get(ti)
 	if err != nil {
 		return fmt.Errorf("error getting signed identity for task %s: %v", task.Name, err)
 	}
 
-	// Derive token for task.
-	req := consul.JWTLoginRequest{
-		JWT:            jwt.JWT,
-		AuthMethodName: consulConfig.TaskIdentityAuthMethod,
-		Meta: map[string]string{
-			"requested_by": fmt.Sprintf("nomad_task_%s", task.Name),
-		},
+	tokenName := widName + "/" + task.Name
+	token := &consulapi.ACLToken{}
+	if err := readACLToken(swi.ACLAccessTokensB64[tokenName], token); err != nil {
+		h.logger.Error("unable to lookup older token", "task", task.Name, "err", err)
 	}
-	token, err := h.getConsulToken(consulConfig.Name, req)
-	if err != nil {
-		return fmt.Errorf("failed to derive Consul token for task %s: %v", task.Name, err)
+
+	// Derive and store token for task if non present yet
+	if token.AccessorID == "" {
+		h.logger.Debug("logging into consul", "name", ti.IdentityName, "type", ti.WorkloadType)
+		req := consul.JWTLoginRequest{
+			JWT:            swi.JWT,
+			AuthMethodName: consulConfig.TaskIdentityAuthMethod,
+			Meta: map[string]string{
+				"requested_by": fmt.Sprintf("nomad_task_%s", task.Name),
+			},
+		}
+
+		token, err = h.getConsulToken(consulConfig.Name, req)
+		if err != nil {
+			return fmt.Errorf("failed to derive Consul token for task %s: %v", task.Name, err)
+		}
+
+		// Store token in client state so it can be reused in case of desconnection
+		if swi.ACLAccessTokensB64 == nil {
+			// Avoid panics for tokens created before this field was added.
+			swi.ACLAccessTokensB64 = map[string]string{}
+		}
+
+		// If this steps fail, we should log and continue allowing the task to start
+		// if by any chance it restarts, it wont be able to reuse the stored token
+		// and it will create a new one.
+		t, err := writeACLToken(token)
+		if err != nil {
+			h.logger.Warn("error processing access token for", "task", task.Name, "error", err)
+		}
+
+		swi.ACLAccessTokensB64[tokenName] = t
+
+		err = h.widmgr.Set(swi)
```

Comment:
Even if we change this to use its own state and not the WI's state, I think we end up potentially making a bunch of database writes here because we can have multiple tokens (ex. one for each task with a `template` block and one for each service). Can we batch these up?